### PR TITLE
[FLINK-24803][Table SQL/Planner] Fix cast BINARY/VARBINARY to STRING

### PR DESCRIPTION
**Pull request by shenzhu**
## What is the purpose of the change

[FLINK-24803](https://issues.apache.org/jira/browse/FLINK-24803): BINARY/VARBINARY should be printed as regular arrays instead of interpreting them in an arbitrary character set as a string.

## Brief change log

  - Update `CastRuleUtils.java` to add `accessField` in order to support accessing object fields
  - Update `CodeGenUtils.scala` to add `arrFieldAccess` in order to support accessing array values through index
  - Update `BinaryToStringCastRule.java` to rewrite the logic and treat binary as arrays

## Verifying this change

This change is already covered by existing tests, such as `CastRulesTest.java`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/BinaryToStringCastRule.java [hunk 1]
```diff
@@ -20,16 +20,18 @@
...
+import static org.apache.flink.table.planner.functions.casting.CastRuleUtils.accessField;
 import static org.apache.flink.table.planner.functions.casting.CastRuleUtils.constructorCall;
+import static org.apache.flink.table.planner.functions.casting.CastRuleUtils.methodCall;
+import static org.apache.flink.table.planner.functions.casting.CastRuleUtils.strLiteral;
 
-/**
- * {@link LogicalTypeFamily#BINARY_STRING} to {@link LogicalTypeFamily#CHARACTER_STRING} cast rule.
- */
-class BinaryToStringCastRule extends AbstractCharacterFamilyTargetRule<byte[]> {
+class BinaryToStringCastRule extends AbstractNullAwareCodeGeneratorCastRule<byte[], String> {
```
Comment (slinkydeveloper) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/BinaryToStringCastRule.java:34]: The generic return value is `StringData`

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/BinaryToStringCastRule.java [hunk 2]
```diff
@@ -41,13 +43,72 @@ private BinaryToStringCastRule() {
...
+    if (!isNull$0) {
+        builder$1.setLength(0);
+        builder$1.append("[");
+        for (int i$2 = 0; i$2 < _myInput.length; i$2++) {
+            if (i$2 != 0) {
+                builder$1.append(", ");
+            }
+        byte element$3 = -1;
+        element$3 = _myInput[i$2];
+        builder$1.append(element$3);
```
Comment (slinkydeveloper) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/BinaryToStringCastRule.java:58]: Perhaps simplify this code without declaring the element variable but just generate `builder$1.append(_myInput[i$2])`

**Comment by matriv** [comment]
@slinkydeveloper @twalthr Should we "guard" this change with the use of the legacyBehaviour flag?

**Comment by twalthr** [comment]
I think we don't need to guard this. The previous behavior makes no sense at all. It can be considered a bug.

**Comment by matriv** [comment]
I'm also +1 to not guard this.

**Comment by matriv** [comment]
@shenzhu Could you please rebase with master and fix the conflicts?

**Comment by matriv** [comment]
@shenzhu Do you have time to finish up this soon, otherwise we would like to take over and finish up the BINARY related Cast issues.

**Comment by matriv** [comment]
@shenzhu I've taken over the PR and made the necessary adjustments to also incorporate padding and trimming.
This is is superseded by: https://github.com/apache/flink/pull/18221

**Comment by shenzhu** [comment]
> #18221

Hey @matriv , so sorry for my late reply! I was on vacation and didn't check Github notifications, thanks so much for your help! I will close this PR.

**Comment by shenzhu** [comment]
replaced by https://github.com/apache/flink/pull/18221

**Comment by matriv** [comment]
@shenzhu , thx for your initial effort! We just wanted to finish up the BINARY related casting, and felt quicker to do the rebasing and apply the adjustments for padding & trimming.

**Pull request closed without merging**